### PR TITLE
Add comment help keyword completion

### LINKED ISSUE #3277: IntelliSense missing in comment blocks
I would have expected completion / IntelliSense for the built-in PowerShell headers in comment help blocks:

```powershell
<#
.SYNOPSIS
    Short description
.DESCRIPTION
    Long description
.EXAMPLE
    PS C:\> <example usage>
    Explanation of what the example does
.INPUTS
    Inputs (if any)
.OUTPUTS
    Output (if any)
.NOTES
    General notes
#>
```

All the supported headers like `.SYNOPSIS` ought to auto-complete (but they don't).

### PR DESCRIPTION
**Pull request by MartinGC94**
# PR Summary
Fixes https://github.com/PowerShell/vscode-powershell/issues/3277 by adding completion for help keywords like "DESCRIPTION" as well as their arguments like the parameter names for "PARAMETER".

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by MartinGC94** [comment]
It seems like my tests are failing because of the carriage returns on mac/Linux but works fine on Windows.  
What's the correct way to fix this? Do I add a `.Replace("`r",'')` to all of the test strings and set my offset accordingly or can I do something to ensure the same line endings on all test platforms?

**Comment by vexx32** [comment]
> It seems like my tests are failing because of the carriage returns on mac/Linux but works fine on Windows.
> What's the correct way to fix this? Do I add a `` .Replace("`r",'') `` to all of the test strings and set my offset accordingly or can I do something to ensure the same line endings on all test platforms?

One approach to this might be to avoid explicit multiline here-strings and construct the test strings like this instead:
```powershell
$TestString = @(
    "line 1"
    "line 2"
    "line 3"
) -join [Environment]::NewLine
```

**Comment by MartinGC94** [comment]
Good idea about using join. I ended up using a character that could easily be removed as a fake cursor, this also makes it easier to tell where the cursor is supposed to be when looking at the test.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5383]:
```suggestion
            FunctionDefinitionAst foundFunction = GetCommentHelpFunctionTarget(context);
            
            ReadOnlyCollection<ParameterAst> foundParameters = null;
            if (foundFunction is not null)
            {
                foundParameters = foundFunction.Parameters ?? foundFunction.Body.ParamBlock?.Parameters;
            }
            else if (context.RelatedAsts[^1] is ScriptBlockAst scriptAst)
            {
                // The helpblock is for a script file
                foundParameters = scriptAst.ParamBlock?.Parameters;
            }
            
            if (foundParameters is null || foundParameters.Count == 0)
            {
                return null;
            }
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: All LINQ usage has now been removed from this PR.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5263]:
```suggestion
                return CompleteCommand(keywordArgument.Value, "*", CommandTypes.All);
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5293]:
```suggestion
                return CompleteFilename(context, containerOnly: false, new HashSet<string>{ ".xml" });
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Agreed. I've changed it to return null because the rest of the completion code that runs after this method checks for null or empty lists.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: This caused issues in situations where it was null. I've added a null check as the first thing before the integer comparison.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5263]: This won't compile for me because of:

"CS0266	Cannot implicitly convert type 'System.Collections.Generic.IEnumerable<System.Management.Automation.CompletionResult>' to 'System.Collections.Generic.List<System.Management.Automation.CompletionResult>'. An explicit conversion exists (are you missing a cast?)

Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5263]: Ah, I see. Adding the new list constructor is fine for now. We should go back through this code and unify the return types and see if we can return null in places where we would otherwise have an empty list.

Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: I've started by fixing my code at least so it will always return a populated list or null.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5213]:
```suggestion
            }

```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5295]:
```suggestion
                return new List<CompletionResult>(CompleteFilename(context, containerOnly: false, (new HashSet<string>() { ".xml" })));
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5351]: Please label the boolean parameter

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5370]:
```suggestion
            }

```

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1
Comment (rjmholt) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1382]: You should write a function to factor this out in every case.

Ideally you could turn these tests into a test cases parameter to `It` with the completions you expect.

The tests should be specific though. Rather than looking at `$rest.CompletionMatches.Count`, the test should specifically expect completion output.

Comment (rjmholt) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:1382]: Example:

https://github.com/PowerShell/PowerShell/blob/6d4420f4b96fefa92b4a03080c1b36e3bb064bf0/test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1#L231-L237

Comment (MartinGC94) [reply test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1]: Thanks, I didn't realize you could make testcases with dynamic names like "<inputStr>".

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5353]:
```suggestion

            // comment based help can apply to a following function definition if it starts within 2 lines
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5379]: Do we really need the null check?
```suggestion
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:4926]:
```suggestion
            if (context.WordToComplete.StartsWith("<#", StringComparison.Ordinal))
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5221]: Why not do it in the previous cycle?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5212]: Sure we can, that was a good idea.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5264]: I haven't tested specifically for FORWARDHELPTARGETNAME but other keywords like PARAMETER and SYNOPSIS do ignore case.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5306]: Maybe? PS isn't localized to my language so I don't know what the standard is but other completions use static English strings like I did here.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5320]: I've wondered that myself, I don't think it's an invisible character because I've tried looking for one and couldn't find it.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs]: Not according to the documentation: The asts for all of the statements represented by this statement block. This property is never null.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5223]: Why should we? I don't think moving that entire code block into the if statement of the previous for loop makes things faster or more readable.

Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5223]: I see your implementation reflects your step-by-step thinking. I have no strong doubts about performance in a completion code but for me putting one condition in one cycle and second condition in second cycle looks strange. I don't ask you to change this now until @rjmholt has doubts about the performance of a code in the PR as a whole.

**Comment by iSazonov** [comment]
@MartinGC94 Please resolve merge conflicts.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5320]: Try to remove all end chars until next line and re-type.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5306]: @rjmholt Thoughts?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5264]: Yep

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5320]: I've tried copying it and removing any visible chars. The only invisible chars left are new lines, carriage returns and spaces (ascii 10,13,32).

**Comment by MartinGC94** [comment]
I don't know how to resolve the merge conflict. Github has a "Resolve conflicts" button but there doesn't seem to be a way to save changes made there. I've tried creating empty space above my newly added tests to make room for the recently added test.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5306]: Yeah that's a good point. @adityapatwardhan do you know if there's a good place for localised descriptions of comment-based help keywords?

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5320]: I suspect the C# TextMate grammar trying to parse this is having trouble, since it's also randomly highlighting keywords in the string. So I doubt the red is meaningful.

The best thing we could do to fix this would be to break this file down to be in smaller component files.

**Comment by iSazonov** [comment]
> there doesn't seem to be a way to save changes made there

Press "Mark as resolved" after you edit.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5223]: The logic here looks fine to me. It's a cheap condition, and it separates the code into a nice `if`/`return` cascade, which I think is very readable (certainly more than embedding a particular case into a loop).

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5231]: Cheaper condition first, since short-circuiting might save us a lookup.

```suggestion
                    if (keyword == lineKeyword || s_commentHelpAllowedDuplicateKeywords.Contains(keyword.Value))
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5306]: So my thinking is that we shouldn't block the PR on this, but before it's merged we should open an issue to track the work to use resources for this.

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5323]: You might also find that this works. I don't think it's preferable to `var` in methods, but it's very convenient for object members.

```suggestion
        private static readonly HashSet<string> s_commentHelpAllowedDuplicateKeywords = new (StringComparer.OrdinalIgnoreCase)
```

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5356]:
```suggestion
            // Comment-based help can apply to a following function definition if it starts within 2 lines
```

**Review by rjmholt — APPROVED** [review]

File: src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs
Comment (MartinGC94) [reply src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs:5306]: I have created an issue here: https://github.com/PowerShell/PowerShell/issues/15450 to clean this up in general for the completion code.

**Pull request merged**
